src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java

```diff
@@ -400,26 +419,42 @@ public final void onEvent(@Nonnull GenericEvent event)
         else if (event instanceof ReconnectedEvent)
             onReconnect((ReconnectedEvent) event);
 
-        Class<ListenerAdapter> handle = ListenerAdapter.class;
         for (Class<?> clazz : ClassWalker.range(event.getClass(), GenericEvent.class))
         {
-            String name = clazz.getSimpleName();
+            if (unresolved.contains(clazz))
+                continue;
+            MethodHandle mh = methods.computeIfAbsent(clazz, ListenerAdapter::findMethod);
+            if (mh == null)
+            {
+                unresolved.add(clazz);
+                continue;
+            }
+
             try
             {
-                name = "on" + name.substring(0, name.length() - "Event".length());
-                Method method = handle.getDeclaredMethod(name, clazz);
-                method.invoke(this, event);
+                mh.invoke(this, event);
             }
-            catch (NoSuchMethodException | IllegalAccessException ignored) {} // this means this is probably a custom event!
-            catch (InvocationTargetException e)
+            catch (Throwable throwable)
             {
-                Throwable cause = e.getCause();
-                if (cause instanceof RuntimeException)
-                    throw (RuntimeException) cause;
-                if (cause instanceof Error)
-                    throw (Error) cause;
-                throw new RuntimeException(cause);
+                if (throwable instanceof RuntimeException)
+                    throw (RuntimeException) throwable;
+                if (throwable instanceof Error)
+                    throw (Error) throwable;
+                throw new IllegalStateException(throwable);
             }
         }
     }
+
+    private static MethodHandle findMethod(Class<?> clazz)
+    {
+        String name = clazz.getSimpleName();
+        MethodType type = MethodType.methodType(Void.TYPE, clazz);
```

Comment:
This is a change, isn't it? This now enforces that we only handle methods that return `void`. Before, any return could be used.

Note, I don't think this is a big deal, I just wanted to call it out.

Reply:
It actually isn't a change since we only checked the methods in ListenerAdapter anyways.

Reply:
Oh that's right. I forgot that it didn't look at the subclass itself.